File: problem2.java
// Time Complexity :O(n log k)
// Space Complexity :O(k)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

import java.util.PriorityQueue;

//  method merges k sorted linked lists efficiently into one sorted list using a priority queue
//  to select the minimum node at each step
public class problem2 {
    class Solution {
        PriorityQueue<ListNode> pq;

        public ListNode mergeKLists(ListNode[] lists) {
            pq = new PriorityQueue<>((a, b) -> a.val - b.val);
            for (ListNode head : lists) {
                if (head != null)
                    pq.add(head);
            }
            ListNode dummy = new ListNode(-1);
            ListNode curr = dummy;
            while (!pq.isEmpty()) {
                ListNode min = pq.poll();
                curr.next = min;
                curr = curr.next;
                if (min.next != null) {
                    pq.add(min.next);
                }
            }
            return dummy.next;
        }
    }
}

File: problem2.cpp
/*
Time complexity: O(nklog(k))
Space complexity: O(k)
Did this code successfully run on Leetcode : Yes
Any problem you faced while coding this : No
*/

class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        priority_queue <pair< int , ListNode* >, vector<pair< int , ListNode*>>, greater<pair< int , ListNode* >>> pq;
        for(auto i : lists) {
            if(i != NULL) {
                pq.push({i->val, i});
            }
        }
        ListNode * head = NULL;
        ListNode * x = head;
        while(!pq.empty()) {
            
            ListNode * y = pq.top().second;
            pq.pop();

            if(head == NULL) {
                x = new ListNode(y->val);
                head = x;
            }else {
                x ->next = new ListNode(y->val);
                x = x->next;
            }
            if(y -> next != NULL) {
                pq.push({y->next->val, y->next});
            }
        }
        return head;
    }
};